Add WebSocket Compression (#22724)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/websockets.md

Never use `Task.Wait`, `Task.Result`, or similar blocking calls to wait for the socket to complete, as that can cause serious threading issues. Always use `await`.

::: moniker range=">= aspnetcore-6.0"

> ### Using Compression

Reviewer: We avoid gerunds in headings because they don't translate well. Can this be? ...

```suggestion
### Compression
```

Reviewer 2: Or `Use compression"
Also, sentence casing.

> [!WARNING]

> > Enabling compression over encrypted connections makes the application subject to CRIME/BREACH type attacks.

Reviewer: If it must be "CRIME/BREACH type attacks," then hyphenate it ("CRIME/BREACH-type"); otherwise ...

```suggestion
> Enabling compression over encrypted connections makes the app subject to CRIME/BREACH attacks.
```

> > If you are sending sensitive information consider not enabling compression, or make sure you use [WebSocketMessageFlags.DisableCompression](TODO) when calling [WebSocket.SendAsync](TODO).

Reviewer: Code-fence API that we'll link later.

```suggestion
> If sending sensitive information, avoid enabling compression or use `WebSocketMessageFlags.DisableCompression` when calling `WebSocket.SendAsync`.
```

> > This applies to both sides of the WebSocket, it's worth noting that the WebSockets API in the browser doesn't have configuration for disabling compression per send. 

Reviewer: ```suggestion
> This applies to both sides of the WebSocket. Note that the WebSockets API in the browser doesn't have configuration for disabling compression per send. 
```

> If compression of messages over WebSockets is desired then the accept code needs to specify that it allows compression as follows:

Reviewer: The suggestion will add a blank line between the sentence and the code block that follows:

```suggestion
If compression of messages over WebSockets is desired, then the accept code must specify that it allows compression as follows:

```

```csharp
using (WebSocket webSocket = await context.WebSockets.AcceptWebSocketAsync(
    new WebSocketAcceptContext() { DangerousEnableCompression = true }))
{
}
````

> Additionally, there are the [ServerMaxWindowBits](TODO) and [DisableServerContextTakeover](TODO) advanced options on [WebSocketAcceptContext](dotnet/api/microsoft.aspnetcore.http.websocketacceptcontext) that control how the compression works.

Reviewer: ```suggestion
Additionally, `ServerMaxWindowBits` and `DisableServerContextTakeover` advanced options of <xref:Microsoft.AspNetCore.Http.WebSocketAcceptContext> control how the compression works.
```

Author: This sounds off for some reason

Reviewer: It's the "of" bit. It could go ....

```suggestion
Additionally, <xref:Microsoft.AspNetCore.Http.WebSocketAcceptContext> advanced options, `ServerMaxWindowBits` and `DisableServerContextTakeover`, control how the compression works.
```

Reviewer: Also, this is a possibility ...

```suggestion
Additionally, `WebSocketAcceptContext.ServerMaxWindowBits` and `WebSocketAcceptContext.DisableServerContextTakeover` control how the compression works.
```

Reviewer 2: > This sounds off for some reason

agreed. Original is fine if they are really advanced, meaning most users won't use them.

Reviewer 2: ```suggestion
[ServerMaxWindowBits](TODO) and [DisableServerContextTakeover](TODO) are advanced options on [WebSocketAcceptContext](dotnet/api/microsoft.aspnetcore.http.websocketacceptcontext) that control how the compression works.
```

Reviewer 2: Additionally is dead wood.

Reviewer: The reason I suggested a change for the original was to avoid the state verb. I think my last suggestion is probably the best of the litter. 🐱

Reviewer: ```suggestion
`WebSocketAcceptContext.ServerMaxWindowBits` and `WebSocketAcceptContext.DisableServerContextTakeover` control how the compression works.
```

... using `?displayProperty=nameWithType` later when these to go API links.

Author: ```suggestion
`WebSocketAcceptContext.ServerMaxWindowBits` and `WebSocketAcceptContext.DisableServerContextTakeover` are advanced options that control how the compression works.
```

> Compression is negotiated between the client and server when first establishing a connection. You can read more about the negotatiation in the [Compression Extensions for WebSocket RFC](https://datatracker.ietf.org/doc/html/rfc7692#section-7)

Reviewer: ```suggestion
Compression is negotiated between the client and server when first establishing a connection. You can read more about the negotiation in the [Compression Extensions for WebSocket RFC](https://datatracker.ietf.org/doc/html/rfc7692#section-7).
```

Author: What do you mean "negotatiation" isn't a word?!

> [!NOTE]

> > If the compression negotiation isn't accepted by either the server or client, the connection will still be established, it will not use compression when sending/receiving messages though.

Reviewer: ```suggestion
> If the compression negotiation isn't accepted by either the server or client, the connection is still established. However, the connection doesn't use compression when sending and receiving messages.
```

Never use `Task.Wait`, `Task.Result`, or similar blocking calls to wait for the socket to complete, as that can cause serious threading issues. Always use `await`.

::: moniker range=">= aspnetcore-6.0"

### Compression

> [!WARNING]

> > Enabling compression over encrypted connections makes the app subject to CRIME/BREACH attacks.

Reviewer 3: can make an app, rather than makes

Reviewer: ```suggestion
> Enabling compression over encrypted connections can make an app subject to CRIME/BREACH attacks.
```
